File: src/main/php/g0001_0100/s0046_permutations/Solution.php
<?php

namespace leetcode\g0001_0100\s0046_permutations;

// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Backtracking
// #Algorithm_I_Day_11_Recursion_Backtracking #Level_2_Day_20_Brute_Force/Backtracking
// #Udemy_Backtracking/Recursion #Top_Interview_150_Backtracking #Big_O_Time_O(n*n!)_Space_O(n+n!)
// #2023_12_09_Time_4_ms_(92.42%)_Space_19.2_MB_(92.42%)

class Solution {
    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    public function permute($nums) {
        if ($nums == null || count($nums) == 0) {
            return array();
        }
        $finalResult = array();
        $this->permuteRecur($nums, $finalResult, array(), array_fill(0, count($nums), false));
        return $finalResult;
    }

    private function permuteRecur($nums, &$finalResult, $currResult, $used) {
        if (count($currResult) == count($nums)) {
            array_push($finalResult, $currResult);
            return;
        }
        for ($i = 0; $i < count($nums); $i++) {
            if ($used[$i]) {
                continue;
            }
            array_push($currResult, $nums[$i]);
            $used[$i] = true;
            $this->permuteRecur($nums, $finalResult, $currResult, $used);
            $used[$i] = false;
            array_pop($currResult);
        }
    }
}

File: src/main/python/g0001_0100/s0046_permutations/Solution.py
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Backtracking
# #Algorithm_I_Day_11_Recursion_Backtracking #Level_2_Day_20_Brute_Force/Backtracking
# #Udemy_Backtracking/Recursion #Top_Interview_150_Backtracking #Big_O_Time_O(n*n!)_Space_O(n+n!)
# #2024_06_08_Time_28_ms_(98.67%)_Space_16.6_MB_(93.71%)

class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        if not nums:
            return []
        
        final_result = []
        self.permute_recur(nums, final_result, [], [False] * len(nums))
        return final_result

    def permute_recur(self, nums: List[int], final_result: List[List[int]], curr_result: List[int], used: List[bool]):
        if len(curr_result) == len(nums):
            final_result.append(curr_result.copy())
            return
        
        for i in range(len(nums)):
            if used[i]:
                continue
            curr_result.append(nums[i])
            used[i] = True
            self.permute_recur(nums, final_result, curr_result, used)
            used[i] = False
            curr_result.pop()
